compiler/rustc_mir/src/monomorphize/collector.rs

```diff
@@ -539,7 +539,7 @@ fn check_recursion_limit<'tcx>(
     // Code that needs to instantiate the same function recursively
     // more than the recursion limit is assumed to be causing an
     // infinite expansion.
-    if !tcx.sess.recursion_limit().value_within_limit(adjusted_recursion_depth) {
+    if !tcx.recursion_limit(()).value_within_limit(adjusted_recursion_depth) {
```

Comment:
We should definitely be able do some caching here. The monomorphization collection already is a monolithic query. Reading the various limits via queries once and then passing them around should be relatively straightforward.